packages/lib-runtime/src/store/coderefs.ts

```diff
@@ -16,6 +16,20 @@ class CodeRefError extends CustomError {
   }
 }
 
+class ExtensionCodeRefsResolutionError extends CustomError {
```

Comment:
Any concern with not exporting this? Although I expect to only interact with these errors through the type guard but the extra properties are not accessible from a TS point of view, right? `e.message` sure, but `e.resolutionErrors` or `e.extension` would be TS invalid if the `map` function is used elsewhere other than directly next to the `isExtensionCodeRefResolutionError`, TS only infers so much for us.

Reply:
Technically, `ExtensionCodeRefsResolutionError` doesn't have to be exported.

Calling `isExtensionCodeRefsResolutionError` narrows down the argument type to ^^ even if it's not imported directly. (So it's basically a transitive type reference.)

In `useResolvedExtensions` hook we do
```ts
const failedPluginNames = _.uniq(
  rejectedReasons
    .filter(isExtensionCodeRefsResolutionError)
    .map((e) => e.extension.pluginName),
);
```
the `filter` part narrows down the type, so in `map` part the `e` argument is implicitly typed as `ExtensionCodeRefsResolutionError`. We can then type-safely access its `e.extension` property. If you control-click on this property in VS Code, it will jump to the corresponding declaration.

Note that all class members have public visibility by default, plus we're also using a short-hand member declaration syntax (called [Parameter Properties](https://www.typescriptlang.org/docs/handbook/2/classes.html#parameter-properties)) in constructors, for example:
```ts
constructor(
  // has no modifiers -> just a plain constructor argument
  foo: string,
  // has modifiers -> becomes "public readonly" class member initialized with given value
  readonly bar: string,
)
```

Reply:
If the code is disassociated from the type guard check, it will not work:

```
array.filter(typeGuardFunction).map(doMappingFunction)

const doMappingFunction (e /* : any */) = {
}
```

But I guess we don't need to solve this now... the type guard is likely next to the call in these situations. Food for thought. 🤔